src/number_theory/legendre_symbol/mul_character.lean

```diff
@@ -0,0 +1,507 @@
+
+lemma coe_to_fun_mul (χ χ' : mul_char R R') : ⇑(χ * χ') = χ * χ' := rfl
+
+@[protected]
+lemma one_mul (χ : mul_char R R') : (1 : mul_char R R') * χ = χ :=
+begin
+  ext a ha,
+  rw [coe_to_fun_mul, pi.mul_apply, one_eval_of_is_unit ha, one_mul],
+end
+
+@[protected]
+lemma mul_one (χ : mul_char R R') : χ * 1 = χ :=
+begin
+  ext a ha,
+  rw [coe_to_fun_mul, pi.mul_apply, one_eval_of_is_unit ha, mul_one],
+end
+
+/-- The inverse of a multiplicative character. We define it as `inverse ∘ χ`. -/
+noncomputable
+def inv (χ : mul_char R R') : mul_char R R' :=
+{ map_nonunit' :=
+  begin
+    intros a ha,
+    simp only [monoid_hom.to_fun_eq_coe, monoid_hom.coe_comp, function.comp_app,
+               monoid_with_zero_hom.to_monoid_hom_coe, monoid_with_zero.coe_inverse, coe_coe,
+               map_nonunit _ ha, ring.inverse_zero],
+  end,
+  ..monoid_with_zero.inverse.to_monoid_hom.comp χ.to_monoid_hom }
+
+noncomputable
+instance has_inv : has_inv (mul_char R R') := ⟨inv⟩
+
+/-- The inverse of a multiplicative character `χ`, applied to `a`, is the inverse of `χ a`. -/
+lemma inv_apply_eq_inv (χ : mul_char R R') (a : R) :
+  χ⁻¹ a = ring.inverse (χ a) :=
+eq.refl $ inv χ a
+
+/-- Variant when the target is a field -/
+lemma inv_apply_eq_inv' {R' : Type v} [field R'] (χ : mul_char R R') (a : R) :
+  χ⁻¹ a = (χ a)⁻¹ :=
+(inv_apply_eq_inv χ a).trans $ ring.inverse_eq_inv (χ a)
+
+/-- When the domain has a zero, we can as well take the inverse first. -/
+lemma inv_apply {R : Type u} [comm_monoid_with_zero R] (χ : mul_char R R') (a : R) :
+  χ⁻¹ a = χ (ring.inverse a) :=
+begin
+  by_cases ha : is_unit a,
+  { rw [inv_apply_eq_inv],
+    have h := is_unit.map χ ha,
+    apply_fun ((*) (χ a)) using is_unit.mul_right_injective h,
+    rw [ring.mul_inverse_cancel _ h, ← map_mul, ring.mul_inverse_cancel _ ha, mul_char.map_one], },
+  { revert ha, nontriviality R, intro ha, -- `nontriviality R` by itself doesn't do it
+    rw [map_nonunit _ ha, ring.inverse_non_unit a ha, mul_char.map_zero χ], },
+end
+
+/-- When the domain is a field, we can use the field inverse instead. -/
+lemma inv_apply' {R : Type u} [field R] (χ : mul_char R R') (a : R) : χ⁻¹ a = χ a⁻¹ :=
+(inv_apply χ a).trans $ congr_arg _ (ring.inverse_eq_inv a)
+
+/-- The product of a character with its inverse is the trivial character. -/
+@[simp]
+lemma inv_mul (χ : mul_char R R') : χ⁻¹ * χ = 1 :=
+begin
+  ext x hx,
+  rw [coe_to_fun_mul, pi.mul_apply, inv_apply_eq_inv,
+      ring.inverse_mul_cancel _ (is_unit.map _ hx), one_eval_of_is_unit hx],
+end
+
+/-- Finally, the commutative group structure on `mul_char R R'`. -/
+noncomputable
+instance comm_group : comm_group (mul_char R R') :=
+{ mul_left_inv := inv_mul,
+  mul_assoc := by { intros χ₁ χ₂ χ₃, ext a, repeat {rw [coe_to_fun_mul]}, rw [mul_assoc], },
+  mul_comm := by { intros χ₁ χ₂, ext a, repeat {rw [coe_to_fun_mul]}, rw [mul_comm], },
+  one_mul := one_mul,
+  mul_one := mul_one,
+  ..has_one,
+  ..has_mul,
+  ..has_inv }
```

Comment:
Also here I want to specify the data-carrying fields explicitly. In addition, `simp` is smart enough to handle `mul_assoc`/`mul_comm` if you ask it:
```suggestion
{ one := 1,
  mul := (*),
  inv := has_inv.inv,
  mul_left_inv := inv_mul,
  mul_assoc := by { intros χ₁ χ₂ χ₃, ext a, simp [mul_assoc], },
  mul_comm := by { intros χ₁ χ₂, ext a, simp [mul_comm], },
  one_mul := one_mul,
  mul_one := mul_one }
```

Reply:
OK.